projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/control/LLVMDispatchBasicBlockNode.java

```diff
@@ -93,11 +98,36 @@ public Object executeGeneric(VirtualFrame frame) {
         int backEdgeCounter = 0;
         outer: while (basicBlockIndex != LLVMBasicBlockNode.RETURN_FROM_FUNCTION) {
             CompilerAsserts.partialEvaluationConstant(basicBlockIndex);
-            LLVMBasicBlockNode bb = bodyNodes[basicBlockIndex];
+
+            if(bodyNodes[basicBlockIndex] instanceof LLVMLoopNode) {
+                LLVMLoopNode loop = (LLVMLoopNode) bodyNodes[basicBlockIndex];
+                loop.execute(frame);
+
+                Integer[] successors = loop.getSuccessors();
+                for(int i = 0; i < successors.length; i++) {
```

Comment:
the iteration should stop at successors.length - 1 and there should be some explicit code to handle the the last successor (see the pattern used for the LLVMSwitchNode below). Otherwise, the graph will look a bit worse.